oracle/src/sender.js

```diff
@@ -106,7 +106,7 @@ async function main({ msg, ackMsg, nackMsg, channel, scheduleForRetry }) {
 
     logger.debug(`Sending ${txArray.length} transactions`)
     await syncForEach(txArray, async job => {
-      const gasLimit = addExtraGas(job.gasEstimate, EXTRA_GAS_PERCENTAGE)
+      const gasLimit = addExtraGas(job.gasEstimate, EXTRA_GAS_PERCENTAGE, MAX_GAS_LIMIT)
```

Comment:
The main problem here, is that we cannot say for sure if the returned value of `eth_estimateGas` includes the amount of gas needed for processing the message or not. For handling that, I made an assumption: if estimated gas is already greater than a message gas limit, I consider it as a final gas limit for message processing; otherwise, I explicitly add a `msgGasLimit` to that value. In case of small, `msgGasLimit`, nothing bad will happen, since we have a high enough multiplier. Let's look on the examples (`bridgeGasLimit`: gas required for bridge contract operation excluding message execution):
* `bridgeGasLimit` = `300'000`, `msgGasLimit` = `150'000`
  - estimateGas may return `300'000` or `450'000 `
  - after using a multiplier we will get something like `1'500'000`/`2'225'000`, which seems fine
* `bridgeGasLimit` = `300'000`, `msgGasLimit` = `500'000`
  - estimateGas may return `300'000` or `800'000`. `300'000` will be transformed into `800'000`
  - after using a multiplier we will get something like `4'000'000`, which also seems fine
* `bridgeGasLimit` = `300'000`, `msgGasLimit` = `2'000'000`
  - estimateGas may return `300'000` or `2'300'000`. `300'000` will be transformed into `2'300'000`
  - after using a multiplier we will get `10'000'000`, since the value exceeds `MAX_GAS_LIMIT`

Maybe we can tune `MAX_GAS_LIMIT` and `EXTRA_GAS_PERCENTAGE` a little bit for some additional accuracy. Or try to rethink of a formula. What numbers for the tx gas limit do you want to achieve in these three cases?

Reply:
The initial introduction of `EXTRA_GAS_PERCENTAGE` was made for the vanilla token bridge. The idea was to preserve the gas if the transaction submitted by the validator eventually became the execution transaction rather than the confirmation transaction. The execution transaction requires more gas for the token minting or invoking the BlockReward contract.

It was reasonable to use a high multiplier there since the transaction to confirm the relay operation consumed relatively small amount of gas, so if we the multiplier is `4` it was still a small number and can be compared with usual transactions. Moreover the validator most probably has enough funds to be allowed to claim submission of such transaction to the chain. 

With AMB the things become a bit more complicated. We cannot just use the same multiplier since the execution transactions will consume the gas that close to `4 * (<bridge operations> + <mediator operations>)` only in the case of the low gas consumption by the mediator. In other cases it will be lower this value significantly.

Let's try to find which components consuming the gas are contained in the bridge operations:
1) bytes transferred in the 'data' filed of the transactions
2) verification of the message and marking it as handled
3) setting fields used by the mediator (like `messageSender`)
4) data transferred to `call` (?) as part of the mediator contract invocation
5) post-actions after the mediator contracts invocation
what else?

I see only _4_ that cannot be estimated in advance for the case when the confirmation transactions are sent by the oracles at the same time. But other components are quite predictable. Even for _4_ we can just to calculate the number of bytes that will be passed to `call` and include this value manually in the estimated gas.

What do you think?

Reply:
Ok, I got you. Then, I propose to do the following:
1. Remove extra gas multipliers for AMB CollectedSignatures and AMB AffirmationRequest watchers.
2. Always add the `msgGasLimit` to the result returned from `eth_gasEstimate`.
3. Add a fixed constant to the final gas limit to cover possible additional gas expenses, (marking message as handled, setting and deleting fields like `messageSender`, `call` opcode cost, other message manipulations like unpacking, validating message version, saving info about call results).
 I have inspected the code in details, in worst case we get the following:
  * For HomeAMB:
    - `800`: SLOAD for source chain id
    - `5000`: SSTORE for marking message as processed
    - memory unpacking and processing is quite cheap, let's say `10000` gas for message unpacking and processing is enough
    - `3 * (20000 + 5000 - 15000) = 30000` for setting and clearing `messageId`, `messageSender`, `messageSourceChainId`
    - `700`: for `CALL` opcode
    - `4 * 20000 = 80000` for storing execution status and 3 failed message fields
    - `375 + 4 * 375 + 32 * 8 = 2131` for emitting an event after message processing
    - In total: `~129000`
  * For ForeignAMB:
    - `3 * 20000 = 60000` for storing 3 failed message fields, if it failed unexpectedly.

I added a `200000` extra gas constant for both watchers, I think it should be enough.

Reply:
Thanks! It looks more logical now. In fact, I believe to use `gasEstimate + (gasEstimate > msgGasLimit ? 0 : msgGasLimit)` is more straightforward in order to not increase the gas limit significantly if the gas consumption by the mediator is closer to its limit. Consider to return it back.

Reply:
I am worry that it can lead to a out-of-gas error. 
If `gasEstimate` is `101000`, `eth_estimateGas` does not consider message execution and `msgGasLimit` is `100000`. Then the final tx gas limit will be `101000 + 200000 = 301000`, however, tx execution can consume up to `101000 + 129000 + 100000 = 330000 > 301000`.

Reply:
Understood.

Let's look on the real numbers.

#### Foreign->Home

A simple contract with a method that consumes quite small amount of gas: https://blockscout.com/poa/sokol/address/0xd1A11bf24794488E3F60408E9E7796b0e351A7ad

**Test 1**, the message is executed successfully
  * Params: `requireToPassMessage('0xd1A11bf24794488E3F60408E9E7796b0e351A7ad', '0xf8a8fd6d', 2000)`
  * [the confirmation tx](https://blockscout.com/poa/sokol/tx/0xb774fa1f151c283e94914955fe20c9db0520a9658f876e481d207ea89959ce4c), `85'890` gas consumed
  * [the execution tx](https://blockscout.com/poa/sokol/tx/0x6692d304bd83eda25e3175a7a70659eef259d3402f10005c5a44acd66bde7a4a), `107'737` gas consumed

**Test 2**, the message is delivered successfully but failed
  * Params: `requireToPassMessage('0xd1A11bf24794488E3F60408E9E7796b0e351A7ad', '0xf8a8fd6d', 500)`
  * [the confirmation tx](https://blockscout.com/poa/sokol/tx/0xdec5fede64a98d0850201df3a232f6d041f114b7cd026dd3569eed4ecf485206), `85'890` gas consumed
  * [the execution tx](https://blockscout.com/poa/sokol/tx/0x44f860b1ff3e65be9bd1595de589b11806b18555484a1f5854193a4f58ff43f4), `151'737` gas consumed

Another contract with a method which assumes small amount of gas but operates with large amount of data: https://blockscout.com/poa/sokol/address/0xF9CEfaf87e411e492c9aE1Ad8f221521E1126bD8

**Test 3**, small amount of data passed to the method, the message is executed successfully
  * Params: `requireToPassMessage('0xF9CEfaf87e411e492c9aE1Ad8f221521E1126bD8', <data>, 2000)`, to look at `data` refer to [the originating tx](https://kovan.etherscan.io/tx/0xb05ea9ff0139e469bb5848467d47376f8c6a2b36a2d0519fd9eba3155b679c5d)
  * [the confirmation tx](https://blockscout.com/poa/sokol/tx/0x664326d7a26b724b96bbd53e83e6397b0a54b8ea7693f6c0f42d27ebec98a267), `86'646` gas consumed
  * [the execution tx](https://blockscout.com/poa/sokol/tx/0x9059346488ac820cace9a81a8099cd8b356c2daeea20c7e0703326912a4e908e), `109'724` gas consumed

**Test 4**, large amount (>13 Kb) of data passed to the method, the message is executed successfully
  * Params: `requireToPassMessage('0xF9CEfaf87e411e492c9aE1Ad8f221521E1126bD8', <data>, 1000000)`, to look at `data` refer to [the originating tx](https://kovan.etherscan.io/tx/0xe8f63d7b860feaa4ea2cf0ad9232fed3f9046aa6844ba60f82b7e7aa5dd678b0)
  * [the confirmation tx](https://blockscout.com/poa/sokol/tx/0xa2e67872b9ebb216ad25614c450e1cb16221c2d2e133603aae8c02f569d4d308), `334'767` gas consumed
  * [the execution tx](https://blockscout.com/poa/sokol/tx/0x2991ed1875df65b732aaaeed742badc178ebe969a2b1be36e071061ae43e7cf3), `874'801` gas consumed

**Test 5**, large amount (>13 Kb) of data passed to the method, **the message is failed to deliver** if the validators are sending the confirmation simultaneously but executed successfully if the last validator delays with the confirmation sending 
  * Params: `requireToPassMessage('0xF9CEfaf87e411e492c9aE1Ad8f221521E1126bD8', <data>, 215620)`, to look at `data` refer to [the originating tx](https://kovan.etherscan.io/tx/0xbb1d0472820ff4cbdfdffd4205709286a82227b5e63ec0ebefbbc03edf4242c6)
  * [the confirmation tx](https://blockscout.com/poa/sokol/tx/0xfc5b6175166c2bfd1587136df21c16edfe2efd7a75dc066fa3e0736a22a43efa), `334'767` gas consumed
  * [the execution tx](https://blockscout.com/poa/sokol/tx/0x129e768bdc7b9b2a12ad68548d7b9a1659752a529759e3b5ed5f5b609bc1e55f), `826,955` gas consumed but it is not enough

The formula to calculate the gas limit on that oracles is:
  1. `gas limit = gasEstimate + (gasEstimate > msgGasLimit ? 0 : msgGasLimit)`
  2. `actual gas limit = gas limit + 500'000`

We see that the execution tx failed for the gas limit `836'357` (`336'357 + 500'000`).

But it is necessary to note that it will fails even for the formula suggested currently in this PR:
  1. `gas limit = gasEstimate + msgGasLimit` (`336'357 + 215'620 = 551'977`)
  2. `actual gas limit = gas limit + 200'000` (`751'977`)

The reason why both approaches fails is that because the formula does not assume that the bridge contract will consume the gas for handling (storing) data that is required to be passed to the mediator code.

So, it seems that the formula must take into account this by somehow, e.g. `actual gas limit = gasEstimate + msgGasLimit + (num of data bytes * X) + 200'000`. Here, in order to decide if the component `num of data bytes * X` is correct, it is necessary to understand how exactly the mediator's data is handled (copied, stored) before the execution will be passed to the mediator's code (probably the component could look more complex, like `num of data bytes * V + num of data bytes * W`). Could you make this research?

As soon as a new model is suggested let's verify it by the tests similar I described above. I also did not performed tests in another direction but it makes sense to verify the model for confirmations sent by the validators in the Foreign chain.

Reply:
I performed some tests with calling `executeAffirmation`, extra gas required for its completion seems to be a quadratic function. This is probably happens due to a memory expansion gas cost in the EVM, which is also a quadratic function.
Here is the data I obtained for identical `executeAffirmation` calls, except that message length was different:

message length | gas used
-|-
0|121122
384|132098
768|143688
1152|155929
1536|168819
1920|182361
2304|196554
2688|211397
3072|226890
3456|243036
3840|259832
4224|277278
4608|295376
4992|314123
5376|333522
5760|353573
6144|374273
6528|395625
6912|417627
7296|440280
7680|463585
8064|487539
8448|512144
8832|537400
9216|563308
9600|589866
9984|617075
10368|644934
10752|673444
11136|702606
11520|732418
11904|762881
12288|793995
12672|825758
13056|858174
13440|891241
13824|924957
14208|959325
14592|994344
14976|1030013
15360|1066332

Using quadratic regression, I have obtained the approximate formula for extra gas required:
`extraGas = 0.0022 * len ** 2 + 27.3352 * len`
For computations in the oracle, I have used slightly bigger coefficients:
`extraGas = 0.0035 * len ** 2 + 40 * len`

Reply:
Cool! Thanks